lib/netext/http.go

```diff
@@ -0,0 +1,460 @@
+		}
+
+		resp.Headers = make(map[string]string, len(res.Header))
+		for k, vs := range res.Header {
+			resp.Headers[k] = strings.Join(vs, ", ")
+		}
+
+		resCookies := res.Cookies()
+		resp.Cookies = make(map[string][]*HTTPCookie, len(resCookies))
+		for _, c := range resCookies {
+			resp.Cookies[c.Name] = append(resp.Cookies[c.Name], &HTTPCookie{
+				Name:     c.Name,
+				Value:    c.Value,
+				Domain:   c.Domain,
+				Path:     c.Path,
+				HTTPOnly: c.HttpOnly,
+				Secure:   c.Secure,
+				MaxAge:   c.MaxAge,
+				Expires:  c.Expires.UnixNano() / 1000000,
+			})
+		}
+	}
+
+	if resErr != nil {
+		// Do *not* log errors about the contex being cancelled.
+		select {
+		case <-ctx.Done():
+		default:
+			state.Logger.WithField("error", resErr).Warn("Request Failed")
+		}
+
+		if preq.Throw {
+			return nil, resErr
+		}
+	}
+
+	return resp, nil
+}
+
+// This should be used instead of setting Error as it will correctly set ErrorCode as well
+func (res *Response) setError(err error) {
+	var errorCode, errorMsg = errorCodeForError(err)
+	res.ErrorCode = int(errorCode)
+	if errorMsg == "" {
+		errorMsg = err.Error()
+	}
+	res.Error = errorMsg
+}
+
+// This should be used instead of setting Error as it will correctly set ErrorCode as well
+func (res *Response) setStatusCode(statusCode int) {
+	res.Status = statusCode
+	if statusCode >= 400 && statusCode < 600 {
+		res.ErrorCode = 1000 + statusCode
+		// TODO: maybe set the res.Error to some custom message
+	}
+}
+
+func (res *Response) setTLSInfo(tlsState *tls.ConnectionState) {
+	tlsInfo, oscp := ParseTLSConnState(tlsState)
+	res.TLSVersion = tlsInfo.Version
+	res.TLSCipherSuite = tlsInfo.CipherSuite
+	res.OCSP = oscp
+}
+
+func mergeCookies(req *http.Request, jar *cookiejar.Jar, reqCookies map[string]*HTTPRequestCookie) map[string][]*HTTPRequestCookie {
+	allCookies := make(map[string][]*HTTPRequestCookie)
+	for _, c := range jar.Cookies(req.URL) {
+		allCookies[c.Name] = append(allCookies[c.Name], &HTTPRequestCookie{Name: c.Name, Value: c.Value})
+	}
+	for key, reqCookie := range reqCookies {
+		if jc := allCookies[key]; jc != nil && reqCookie.Replace {
+			allCookies[key] = []*HTTPRequestCookie{{Name: key, Value: reqCookie.Value}}
+		} else {
+			allCookies[key] = append(allCookies[key], &HTTPRequestCookie{Name: key, Value: reqCookie.Value})
+		}
+	}
+	return allCookies
+}
+
+func setRequestCookies(req *http.Request, reqCookies map[string][]*HTTPRequestCookie) {
+	for _, cookies := range reqCookies {
+		for _, c := range cookies {
+			req.AddCookie(&http.Cookie{Name: c.Name, Value: c.Value})
+		}
+	}
+}
+
+func debugRequest(state *lib.State, req *http.Request, description string) {
+	if state.Options.HttpDebug.String != "" {
+		dump, err := httputil.DumpRequestOut(req, state.Options.HttpDebug.String == "full")
+		if err != nil {
+			log.Fatal(err)
+		}
+		logDump(description, dump)
+	}
+}
+
+// ResponseType is used in the request to specify how the response body should be treated
```

Comment:
I think it will be a bit better to have the request and response parts in different files, just to split things up a bit.